src/kmeans.jl

```diff
@@ -18,35 +18,33 @@ const _kmeans_default_maxiter = 100
 const _kmeans_default_tol = 1.0e-6
 const _kmeans_default_display = :none
 
-function kmeans!(X::Matrix{T}, centers::Matrix{T};
-                 weights=nothing,
-                 maxiter::Integer=_kmeans_default_maxiter,
+function kmeans!(X::AbstractMatrix{<:Real}, centers::AbstractMatrix{<:Real};
+                 weights=nothing, maxiter::Integer=_kmeans_default_maxiter,
                  tol::Real=_kmeans_default_tol,
                  display::Symbol=_kmeans_default_display,
-                 distance::SemiMetric=SqEuclidean()) where T<:AbstractFloat
+                 distance::SemiMetric=SqEuclidean())
 
     m, n = size(X)
     m2, k = size(centers)
     m == m2 || throw(DimensionMismatch("Inconsistent array dimensions."))
     (2 <= k < n) || error("k must have 2 <= k < n.")
 
-    assignments = zeros(Int, n)
-    costs = zeros(T, n)
+    assignments = Vector{Int}(undef, n)
+    costs = Vector{Float64}(undef, n)
     counts = Vector{Int}(undef, k)
     cweights = Vector{Float64}(undef, k)
 
-    _kmeans!(X, conv_weights(T, n, weights), centers,
-             assignments, costs, counts, cweights,
-             round(Int, maxiter), tol, display_level(display), distance)
+    _kmeans!(X, conv_weights(Float64, n, weights), centers, assignments,
```

Comment:
`T` was actually needed in this function, otherwise it can become type unstable.

Reply:
With apologies, I don't really understand this comment and more importantly I'm not sure what to do.
Reverting to the previous doesn't seem right to me as it imposes the type of `X` and `centers` to be the same which I don't think is correct. 
I've changed the `Float64` in the `conv_weights` to `eltype(X)` is that ok?

Reply:
`eltype(X)` is fine. You could also have kept `T`, just not using it for `centers`.

However, please check that the code (this function and others down the line) is type stable when `X` and `centers` have different eltypes: it may have been written with the assumption that they are equal.

Reply:
with hindsight, I'm not happy with this conversion. If the `X` are say, integers and the weights are float, the weights should stay floats. So it should be something like `conv_weights(promote_type(eltype(X), Float64), n, weights)`

Reply:
Good catch. But `promote_type(eltype(X), Float64)` will almost always return `Float64`. Better use `float(eltype(X))`. Or maybe better, use `eltype(centers)`, which cannot be integers, so that the user can choose the most appropriate type.

Reply:
> `T` was actually needed in this function, otherwise it can become type unstable.

Are the relevant tests already there (`@inferred` `typeof() ==` etc) or could they be added? Otherwise finding the right types is a little bit like wandering in the dark.

Reply:
I don't think so but maybe that could be for another PR also related to #140 so it's not just for `kmeans` that such checks would be needed...